stan/math/rev/core/scoped_chainablestack.hpp

```diff
@@ -0,0 +1,50 @@
+#ifndef STAN_MATH_REV_CORE_SCOPED_CHAINABLESTACK_HPP
+#define STAN_MATH_REV_CORE_SCOPED_CHAINABLESTACK_HPP
+
+#include <stan/math/rev/core/chainablestack.hpp>
+#include <stan/math/rev/core/start_nested.hpp>
+#include <stan/math/rev/core/recover_memory_nested.hpp>
+
+#include <vector>
+
+namespace stan {
+namespace math {
+
+class ScopedChainableStack {
+  typedef ChainableStack::AutodiffStackStorage chainablestack_t;
+  chainablestack_t local_stack_;
+
+  using stack_queue_t = std::vector<chainablestack_t*>;
+  stack_queue_t stack_queue_;
+
+  struct activate_scope {
+    ScopedChainableStack& scoped_stack_;
+
+    explicit activate_scope(ScopedChainableStack& scoped_stack)
+        : scoped_stack_(scoped_stack) {
+      scoped_stack_.stack_queue_.push_back(ChainableStack::instance_);
+      ChainableStack::instance_ = &scoped_stack_.local_stack_;
+    }
+
+    ~activate_scope() {
+      ChainableStack::instance_ = scoped_stack_.stack_queue_.back();
+      scoped_stack_.stack_queue_.pop_back();
+    }
+  };
+
+ public:
+  ScopedChainableStack() {}
+
+  // execute in the current thread a nullary function and write the AD
+  // tape to local_stack_ of this instance
+  template <typename F>
+  F execute(F f) {
+    activate_scope active_scope(*this);
+    f();
+    return f;
+  }
```

Comment:
Do we need to return F? Or could we just do

```suggestion
  auto execute(F&& f) {
    activate_scope active_scope(*this);
    return std::forward<F>(f)();
  }
```

Because then in the constructor for the reducer we can call this in the initialization list (I'll comment on that below)